pkg/provider/loadbalancer/securitygroup/securitygroup.go

```diff
@@ -303,6 +301,72 @@ func (helper *RuleHelper) AddRuleForDenyAll(dstAddresses []netip.Addr) error {
 	return nil
 }
 
+// RemoveDestinationFromRules removes the given destination addresses from rules that match the given protocol and ports is in the retainDstPorts list.
+// It may add a new rule if the original rule needs to be split.
+func (helper *RuleHelper) RemoveDestinationFromRules(
+	protocol network.SecurityRuleProtocol,
+	dstPrefixes []string,
+	retainDstPorts []int32,
+) error {
+	logger := helper.logger.WithName("RemoveDestinationFromRules").WithValues("protocol", protocol, "num-dst-prefixes", len(dstPrefixes))
+	logger.V(10).Info("Cleaning destination from SecurityGroup")
+
+	for _, rule := range helper.rules {
+		if rule.Protocol != protocol {
+			continue
+		}
+
+		if err := helper.removeDestinationFromRule(rule, dstPrefixes, retainDstPorts); err != nil {
+			logger.Error(err, "Failed to remove destination from rule", "rule-name", *rule.Name)
+		}
+	}
+
+	return nil
+}
+
+func (helper *RuleHelper) removeDestinationFromRule(rule *network.SecurityRule, prefixes []string, retainDstPorts []int32) error {
+	logger := helper.logger.WithName("removeDestinationFromRule").
+		WithValues("security-rule-name", rule.Name)
+	currentPorts, err := ListDestinationPortRanges(rule)
+	if err != nil {
+		// Skip the rule with invalid destination port ranges.
+		// NOTE: cloud-provider would not create allow rules with `*` or `4000-5000` as destination port ranges.
+		return nil
+	}
+
+	var (
+		prefixIndex     = fnutil.IndexSet(prefixes) // Used to check whether the prefix should be removed.
+		currentPrefixes = ListDestinationPrefixes(rule)
+
+		expectedPrefixes = fnutil.RemoveIf(func(p string) bool { return prefixIndex[p] }, currentPrefixes) // The prefixes to keep.
+		targetPrefixes   = fnutil.Intersection(currentPrefixes, prefixes)                                  // The prefixes to remove.
+		expectedPorts    = fnutil.Intersection(currentPorts, retainDstPorts)                               // The ports to keep.
+	)
+
+	// assert: len(targetPrefixes) + len(expectedPrefixes) == len(currentPrefixes)
+	// assert: len(targetPorts) + len(expectedPorts) == len(currentPorts)
+
+	if len(targetPrefixes) == 0 || len(currentPorts) == len(expectedPorts) {
+		return nil
+	}
+
+	rule.DestinationAddressPrefix = nil
+	rule.DestinationAddressPrefixes = ptr.To(NormalizeSecurityRuleAddressPrefixes(expectedPrefixes))
+
+	if len(expectedPorts) == 0 {
```

Comment:
it seems this could be merged into L349, but shouldn't be merged there. Could you add some comments on the rule update steps (to avoid someone breaks the logic in the future)?